### ISSUE TITLE
writeGlobalFIle() not returning jobStoreFileID

### ISSUE THREAD

**Issue by arkal** [BD2KGenomics/toil#469]
writeGlobalFile should be returning a jobStoreFileID that can be passed on to readGlobalFile for retrieval. I've tested this and wGF appears to be returning the value of localFileName instead. Code to follow...

**Comment by arkal** [comment]
test code:
```
#!/usr/bin/env python2.7
"""
Author : Arjun Arkal Rao
Affiliation : UCSC BME, UCSC Genomics Institute
File : test/unit/test_writeGlobalFileError.py

"""
from __future__ import division, print_function
from toil.job import Job

import argparse
import aux_scripts as aux
import os

class writeGlobalFileTest():
    ''' This is to test whether writeGlobalFile returns a jobStoreFileID or no'''
    def __init__(self):
        ''' Setting up for the test '''
        #  Set up input and test directories
        self.input_dir = os.path.abspath('/data1/toil_test_inputs')
        self.test_dir = os.path.abspath('/data1/toil_test_io_phlat')
        #  initialize options. Input files are
        #  /data1/toil_test_inputs/rna_SRR2099838_1.fastq
        #  /data1/toil_test_inputs/rna_SRR2099838_2.fastq
        self.fastqs = {'normal_dna': ['/'.join([self.input_dir,
                                                'rna_SRR2099838_1.fastq']),
                                      '/'.join([self.input_dir,
                                                'rna_SRR2099838_2.fastq'])]}

    def test_wgf_output(self):
        ''' To test whether run_phlat works'''
        parser = argparse.ArgumentParser()
        parser.add_argument('--dummy', dest='dummy', default='dummy')
        Job.Runner.addToilOptions(parser)
        params = parser.parse_args(['--workDir', '/data1/temp_dir', 
                                    '/data1/toil_tempdir'])
        #  Write files into file store. self.fastqs should be updated to hold
        #  jobStoreFileIDs
        F = Job.wrapJobFn(aux.write_to_filestore, self.fastqs)
        #  Print the jobStoreFileIDs
        G = Job.wrapJobFn(aux.print_filepaths, self.fastqs)
        #  get the files corresponding to jobStoreFileIDs.  self.fastqs will be
        #  updated to hold the paths returned by readGlobalFile
        H = Job.wrapJobFn(aux.get_files_from_filestore, self.fastqs)
        #  Print the filepaths
        I = Job.wrapJobFn(aux.print_filepaths, self.fastqs)
        F.addChild(G)
        G.addChild(H)
        H.addChild(I)
        Job.Runner.startToil(F, params)


if __name__ == '__main__':
    A = writeGlobalFileTest()
    A.test_wgf_output()
```

aux.py contains
```
def write_to_filestore(job, files):
    '''
    This will write a dict of files to the write_to_filestore.
    The input FILES object is expected to be a dict of
    key = filename or sample_type
    value = filepath if key == filename
            list of filepaths if key == sample_type
    Eg. {'SRR123.fastq': '/data/SRR123.fastq',
         'sample1': ['/data/s_1.fastq', '/data/s_2.fastq']}

    FILES will be updated to
    key = filename or sample_type
    value = jobStoreFileID if key == filename
            list of jobStoreFileIDs if key == sample_type
    '''
    for name in files.keys():
        #  if list, replace the items with the filestoreID for each item 
        #  obtained from writeGlobalFile
        if isinstance(files[name], list):
            files[name] = [job.fileStore.writeGlobalFile(x) for x in files[name]]
        else:
            files[name] = job.fileStore.writeGlobalFile(files[name])


def print_filepaths(job, infiles):
    '''
    This will write a print the values of the items in INFILES. INFILES is a
    dict of
    key = filename or sample_type
    value = filepath if key == filename
            list of filepaths if key == sample_type
    Eg. {'SRR123.fastq': '/data/SRR123.fastq',
         'sample1': ['/data/s_1.fastq', '/data/s_2.fastq']}
    '''
    with open('/data1/temp_dir/cachepaths.txt', 'a') as cfile:
        for name in infiles.keys():
            #  If list, print each member of the list
            if isinstance(infiles[name], list):
                for item in infiles[name]:
                    print(item, file=cfile)
            #  else print the value for the key
            else:
                print(infiles[name], file=cfile)


def get_files_from_filestore(job, files):
    """
    This is adapted from John Vivian's return_input_paths from the RNA-Seq
    pipeline.

    Returns the paths of files from the FileStore if they are not present.
    files is a dict with
        key = filename or sample_type
        value = filepath if key == filename
                list of filepaths if key == sample_type
    Eg. {'SRR123.fastq': '/data/SRR123.fastq',
         'sample1': ['/data/s_1.fastq', '/data/s_2.fastq']}
    """
    for name in files.keys():
        if isinstance(files[name], list):
            files[name] = [job.fileStore.readGlobalFile(x) for x in files[name]]
        else:
            files[name] = job.fileStore.readGlobalFile(files[name])
```

TL;DR summary - 
1. Take files and put them into the file store.
2. Print the returned jobStoreFileIDs for both files
3. Retrieve the files from the jobstore
4. Print the returned filepaths for both passed  jobStoreFileID

EXPECTED OUTPUT (/data1/temp_dir/cachepaths.txt
```
/data1/toil_tempdir/tmp/X/g/tmpXXXX.tmp
/data1/toil_tempdir/tmp/Y/g/tmpYYYY.tmp
/data1/temp_dir/localtempdir/tmpXXXX.tmp
/data1/temp_dir/localtempdir/tmpYYYY.tmp
```
Actually, I don't know what to expect in the last 2 lines with the new readGlobalFile and I was running this test to find out what i would get.

SEEN OUTPUT for small (~300KB each) and empty files
```
/data1/toil_test_inputs/rna_SRR2099838_1.fastq
/data1/toil_test_inputs/rna_SRR2099838_2.fastq
/data1/toil_test_inputs/rna_SRR2099838_1.fastq
/data1/toil_test_inputs/rna_SRR2099838_2.fastq
```
SEEN OUTPUT for large files (4.5GB each)
```
/data1/toil_test_inputs/rna_SRR2099838_1.fastq
/data1/toil_test_inputs/rna_SRR2099838_2.fastq
```
At which point the program crashes with this error
```
[centos@arjun-fake-podk toil_tests]$ python test_wgf.py 
No handlers could be found for logger "toil.resource"
INFO:toil.lib.bioio:Logging set at level: INFO
INFO:toil.common:Using the single machine batch system
INFO:toil.jobStores.fileJobStore:Jobstore directory is: /data1/toil_tempdir
WARNING:toil.batchSystems.singleMachine:Limiting maxCores to CPU count of system (8).
INFO:toil.batchSystems.singleMachine:Setting up the thread pool with 80 workers, given a minimum CPU fraction of 0.100000 and a maximum CPU value of 8.
INFO:toil.common:Written the environment for the jobs to the environment file
WARNING:toil.resource:Can't globalize module ModuleDescriptor(dirPath='/usr/local/lib/python2.7/site-packages', name='toil.job', extension='.pyc').
WARNING:toil.resource:Can't globalize module ModuleDescriptor(dirPath='/usr/local/lib/python2.7/site-packages', name='toil.job', extension='.pyc').
WARNING:toil.resource:Can't globalize module ModuleDescriptor(dirPath='/usr/local/lib/python2.7/site-packages', name='toil.job', extension='.pyc').
WARNING:toil.resource:Can't globalize module ModuleDescriptor(dirPath='/usr/local/lib/python2.7/site-packages', name='toil.job', extension='.pyc').
INFO:toil.leader:Checked batch system has no running jobs and no updated jobs
INFO:toil.leader:Found 1 jobs to start and 0 jobs with successors to run
INFO:toil.leader:Starting the main loop
INFO:toil.batchSystems.singleMachine:Executing command: '/usr/local/bin/python -E /usr/local/lib/python2.7/site-packages/toil/worker.py /data1/toil_tempdir 3/6/job3op19W'.
Exception in thread Thread-13:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 398, in asyncWrite
    raise RuntimeError("The termination flag is set, exiting")
RuntimeError: The termination flag is set, exiting

Exception in thread Thread-14:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 398, in asyncWrite
    raise RuntimeError("The termination flag is set, exiting")
RuntimeError: The termination flag is set, exiting

Exception RuntimeError: RuntimeError('cannot join current thread',) in <bound method FileStore.__del__ of <toil.job.FileStore object at 0x7f2196494d50>> ignored
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 569, in asyncUpdate
    raise RuntimeError("The termination flag is set, exiting before update")
RuntimeError: The termination flag is set, exiting before update

Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 569, in asyncUpdate
    raise RuntimeError("The termination flag is set, exiting before update")
RuntimeError: The termination flag is set, exiting before update

Exception in thread Thread-9:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 569, in asyncUpdate
    raise RuntimeError("The termination flag is set, exiting before update")
RuntimeError: The termination flag is set, exiting before update

Exception in thread Thread-12:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 569, in asyncUpdate
    raise RuntimeError("The termination flag is set, exiting before update")
RuntimeError: The termination flag is set, exiting before update

WARNING:toil.leader:The jobWrapper seems to have left a log file, indicating failure: 3/6/job3op19W
WARNING:toil.leader:Reporting file: 3/6/job3op19W
WARNING:toil.leader:3/6/job3op19W:	---TOIL WORKER OUTPUT LOG---
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.resource:Can't find resource for leader path '/data/test_docker/toil_tests/aux_scripts.pyc'
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.resource:Can't localize module ModuleDescriptor(dirPath='/data/test_docker/toil_tests', name='aux_scripts', extension='.pyc')
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.resource:Can't find resource for leader path '/data/test_docker/toil_tests/aux_scripts.pyc'
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.resource:Can't localize module ModuleDescriptor(dirPath='/data/test_docker/toil_tests', name='aux_scripts', extension='.pyc')
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.resource:Can't find resource for leader path '/data/test_docker/toil_tests/aux_scripts.pyc'
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.resource:Can't localize module ModuleDescriptor(dirPath='/data/test_docker/toil_tests', name='aux_scripts', extension='.pyc')
WARNING:toil.leader:3/6/job3op19W:	Traceback (most recent call last):
WARNING:toil.leader:3/6/job3op19W:	  File "/usr/local/lib/python2.7/site-packages/toil/worker.py", line 284, in main
WARNING:toil.leader:3/6/job3op19W:	    fileStore=fileStore)
WARNING:toil.leader:3/6/job3op19W:	  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 1072, in _execute
WARNING:toil.leader:3/6/job3op19W:	    returnValues = self.run(fileStore)
WARNING:toil.leader:3/6/job3op19W:	  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 1186, in run
WARNING:toil.leader:3/6/job3op19W:	    rValue = userFunction(*((self,) + tuple(self._args)), **self._kwargs)
WARNING:toil.leader:3/6/job3op19W:	  File "/data/test_docker/toil_tests/aux_scripts.py", line 74, in get_files_from_filestore
WARNING:toil.leader:3/6/job3op19W:	    files[name] = [job.fileStore.readGlobalFile(x) for x in files[name]]
WARNING:toil.leader:3/6/job3op19W:	  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 497, in readGlobalFile
WARNING:toil.leader:3/6/job3op19W:	    self.jobStore.readFile(fileStoreID, localFilePath)
WARNING:toil.leader:3/6/job3op19W:	  File "/usr/local/lib/python2.7/site-packages/toil/jobStores/fileJobStore.py", line 166, in readFile
WARNING:toil.leader:3/6/job3op19W:	    self._checkJobStoreFileID(jobStoreFileID)
WARNING:toil.leader:3/6/job3op19W:	  File "/usr/local/lib/python2.7/site-packages/toil/jobStores/fileJobStore.py", line 279, in _checkJobStoreFileID
WARNING:toil.leader:3/6/job3op19W:	    raise NoSuchFileException("File %s does not exist in jobStore" % jobStoreFileID)
WARNING:toil.leader:3/6/job3op19W:	NoSuchFileException: The file 'File /data1/toil_test_inputs/rna_SRR2099838_1.fastq does not exist in jobStore' does not exist
WARNING:toil.leader:3/6/job3op19W:	Exiting the worker because of a failed jobWrapper on host arjun-fake-podk.novalocal
WARNING:toil.leader:3/6/job3op19W:	ERROR:__main__:Exiting the worker because of a failed jobWrapper on host arjun-fake-podk.novalocal
WARNING:toil.leader:3/6/job3op19W:	WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 3/6/job3op19W to 0
WARNING:toil.leader:Job: 3/6/job3op19W is completely failed
INFO:toil.leader:Only failed jobs and their dependents (1 total) are remaining, so exiting.
INFO:toil.leader:Finished the main loop
INFO:toil.leader:Waiting for stats and logging collator process to finish
INFO:toil.leader:Stats/logging finished collating in 0.348111152649 seconds
Traceback (most recent call last):
  File "test_wgf.py", line 55, in <module>
    A.test_wgf_output()
  File "test_wgf.py", line 50, in test_wgf_output
    Job.Runner.startToil(F, params)
  File "/usr/local/lib/python2.7/site-packages/toil/job.py", line 362, in startToil
    return mainLoop(config, batchSystem, jobStore, rootJob)
  File "/usr/local/lib/python2.7/site-packages/toil/leader.py", line 505, in mainLoop
    raise FailedJobsException( config.jobStore, totalFailedJobs )
toil.leader.FailedJobsException: The job store '/data1/toil_tempdir' contains 1 failed jobs
```

So it looks like writeGlobalFile is not returning  a jobStoreFileID.

**Comment by joelarmstrong** [comment]
Hey Arjun,

I think the problem here is that, in jobTree/toil semantics, all the parameters to a job get saved *at job creation time* and not at job run time. Functions like `write_to_filestore` won't work, because they modify a reference that's been passed. That self.fastqs reference is actually to pickled data and isn't, at job run time, the same reference that is passed to the other functions. Any modified data must be explicitly returned, or passed to children created dynamically.

**Comment by arkal** [comment]
Hi Joel, I just caught that! Thanks for the post though.
